backend/asmgen.ml

```diff
@@ -285,7 +285,8 @@ let compile_unit ~output_prefix ~asm_filename ~keep_asm ~obj_filename gen =
          ~exceptionally:(fun () ->
              if create_asm && not keep_asm then remove_file asm_filename);
        if should_emit () then begin
-         Emitaux.reduce_heap_size ~reset:(fun () ->
+         if may_reduce_heap then
+           Emitaux.reduce_heap_size ~reset:(fun () ->
             reset ();
             Typemod.reset ();
             Emitaux.reset ();
```

Comment:
I've just noticed that Emitaux.reset is redundant (it's called by Emitaux.reduce_heap_size anyway).